File: homework2/Saytam_Mishra/q8_isBST.java
package homework2.Saytam_Mishra;
// Depth-first traversal (Pre-order)
// Time Complexity - O(n)
// Space Complexity - O(n), O(logn) for perfectly balanced tree

public class q8_isBST {
    static class TreeNode {
        TreeNode left;
        TreeNode right;
        int val;

        public TreeNode(int val) {
            this.left = null;
            this.right = null;
            this.val = val;
        }
    }

    public static boolean isBST(TreeNode root, Integer min, Integer max) {
        if(root == null) return true;

        if(min != null && min >= root.val) {
            return false;
        }
        if(max != null && max <= root.val) {
            return false;
        }

        return isBST(root.left, min, root.val) && isBST(root.right, root.val, max);
    }

    public static void main(String[] args) {
        // Test Case 1
        TreeNode root1 = new TreeNode(10);
        root1.left = new TreeNode(8);
        root1.right = new TreeNode(16);
        root1.left.right = new TreeNode(9);
        root1.right.left = new TreeNode(13);
        root1.right.right = new TreeNode(17);
        root1.right.right.right = new TreeNode(20);
        System.out.println("Is Tree 1 a BST: " + isBST(root1, null, null));

        // Test Case 2
        TreeNode root2 = new TreeNode(10);
        root2.left = new TreeNode(8);
        root2.right = new TreeNode(16);
        root2.left.right = new TreeNode(9);
        root2.right.left = new TreeNode(13);
        root2.right.right = new TreeNode(17);
        root2.right.right.right = new TreeNode(15);
        System.out.println("Is Tree 2 a BST: " + isBST(root2, null, null));
    }
}

// Time spent - 25m 34s

File: homework2/Yash/Part 2/Q8_IsBST.cpp
// Q8: Is BST, Technique Used: Search binary search tree technique (min/max recursion)
// Time Complexity: O(n), Space Complexity: O(n) , Time Taken : 32 minutes 

#include <iostream>
using namespace std;

struct Node {
    int val;
    Node* left;
    Node* right;

    Node(int v) {
        val = v;
        left = nullptr;
        right = nullptr;
    }
};


bool isBSTHelper(Node* root, int minVal, int maxVal) {
    if (root == nullptr) {
        return true;
    }

    if (root->val <= minVal || root->val >= maxVal) {
        return false;
    }

    bool leftValid = isBSTHelper(root->left, minVal, root->val);
    bool rightValid = isBSTHelper(root->right, root->val, maxVal);

    return leftValid && rightValid;
}


bool isBST(Node* root) {
    return isBSTHelper(root, -2147483648, 2147483647);
}

// Simple preorder print 
void printPre(Node* root) {
    if (root == nullptr) {
        return;
    }
    cout << root->val << " ";
    printPre(root->left);
    printPre(root->right);
}


int main() {
    cout << "Testing IsBST\n";

    // Example 1: Valid BST
    //        10
    //       /  \
    //      8    16
    //       \   / \
    //       9 13  17
    //              \
    //              20         
    Node* root1 = new Node(10);
    root1->left = new Node(8);
    root1->right = new Node(16);

    
    root1->left->right = new Node(9);

    root1->right->left = new Node(13);
    root1->right->right = new Node(17);

    root1->right->right->right = new Node(20);

    cout << "Tree 1 IsBST: " << isBST(root1) << "\n";

    // Example 2: Invalid BST
    //       5
    //      / \
    //     1   4
    //        / \
    //       3   6

    Node* root2 = new Node(5);
    root2->left = new Node(1);
    root2->right = new Node(4);
    root2->right->left = new Node(3);
    root2->right->right = new Node(6);

    cout << "Tree 2 IsBST: " << isBST(root2) << "\n";

    return 0;
}
